kloppy/domain/services/state_builder/lineup.py

```diff
@@ -0,0 +1,58 @@
+from dataclasses import dataclass
+from typing import Set
+
+from kloppy.domain import (
+    Event,
+    EventDataset,
+    Player,
+    SubstitutionEvent,
+    PlayerOffEvent,
+    PlayerOnEvent,
+    CardEvent,
+    CardType,
+    Provider,
+)
+from . import StateBuilder
+
+
+@dataclass
+class Lineup:
+    players: Set[Player]
+
+
+class LineupStateBuilder(StateBuilder):
+    def initial_state(self, dataset: EventDataset) -> Lineup:
+        if dataset.metadata.provider != Provider.STATSBOMB:
+            raise Exception(
+                "Lineup state can only be applied to statsbomb data"
+            )
+
+        return Lineup(
+            players=(
+                set(
+                    player
+                    for player in dataset.metadata.teams[0].players
+                    if player.starting
+                )
+                | set(
+                    player
+                    for player in dataset.metadata.teams[1].players
+                    if player.starting
+                )
+            )
+        )
+
+    def reduce(self, state: Lineup, event: Event) -> Lineup:
+        if isinstance(event, SubstitutionEvent):
+            state = Lineup(
+                players=state.players - {event.player}
+                | {event.replacement_player}
+            )
+        elif isinstance(event, PlayerOffEvent):
```

Comment:
My apologies for not being familiar with the dataset formats, but is the differentiation between SubstitutionEvent and PlayerOn/Off because different providers have different ways of representing substitutions?

Reply:
Substitution is the official when a player is substituted. The PlayerOff/PlayerOn are the ones when a player leaves the pitch for any other reason, for injury treatment for example.

Reply:
I see, thank you!